models/inventory/tests/testStockTransfer.spec.ts

```diff
@@ -585,4 +590,211 @@ test('Create Shipment from manually set Back Ref', async (t) => {
+    ModelNameEnum.AccountingLedgerEntry,
+    {
+      fields: ['name', 'account', 'credit', 'debit'],
+      filters: { referenceName: creditNoteDoc.name! },
+    }
+  );
+
+  for (const ale of creditNoteAles) {
+    if (ale.account === 'Sales') {
+      t.equal(
+        ale.credit,
+        '0.00000000000',
+        `Credit Note *not credited to ${ale.account}`
+      );
+      t.equal(
+        ale.debit,
+        '-300.00000000000',
+        `Credit Note debited from ${ale.account}`
+      );
+    }
+
+    if (ale.account === 'Debtors') {
+      t.equal(
+        ale.debit,
+        '0.00000000000',
+        `Credit Note *not debited from ${ale.account}`
+      );
+      t.equal(
+        ale.credit,
+        '-300.00000000000',
+        `Credit Note credited to ${ale.account}`
+      );
+    }
+  }
+
+  const invoiceOutstanding = Object.values(
+    await fyo.db.get(
+      ModelNameEnum.SalesInvoice,
+      sinv.name!,
+      'outstandingAmount'
+    )
+  )[0] as Money;
+
+  t.true(invoiceOutstanding.isZero(), 'Sales Invoice outstanding is Zero');
+
+  await assertThrows(
+    async () => await sinv.cancel(),
+    'can not cancel a SINV when a credit note is created against it'
+  );
+
+  await assertDoesNotThrow(
+    async () => await creditNoteDoc.cancel(),
+    'credit note cancelled'
+  );
+
+  const updatedInvoiceOutstanding = Object.values(
+    await fyo.db.get(
+      ModelNameEnum.SalesInvoice,
+      sinv.name!,
+      'outstandingAmount'
+    )
+  )[0] as Money;
+
+  t.ok(
+    updatedInvoiceOutstanding.eq(sinv.outstandingAmount as Money),
+    'outstandingAmount updated after credit note is cancelled'
+  );
+});
+
+test('Create Purchase Invoice then create Debit Note against it', async (t) => {
+  const rate = testDocs.Item[item].rate as number;
+  const pinv = fyo.doc.getNewDoc(
+    ModelNameEnum.PurchaseInvoice
+  ) as PurchaseInvoice;
+
+  const date = new Date('2022-01-04');
+  await pinv.set({
+    date,
+    party,
+    account: 'Creditors',
+  });
+  await pinv.append('items', { item, quantity: 3, rate });
+  await pinv.sync();
+  await pinv.submit();
+
+  t.equal(pinv.name, 'PINV-1002', 'PINV name matches');
+
+  const rawDebitNote = (await createReturnDoc(
+    pinv.getValidDict(true, true),
+    ModelNameEnum.PurchaseInvoice,
+    fyo
+  )) as DocValueMap;
+
+  const debitNoteDoc = fyo.doc.getNewDoc(
+    ModelNameEnum.PurchaseInvoice,
+    rawDebitNote,
+    true
+  ) as PurchaseInvoice;
+
+  await debitNoteDoc.sync();
+  await debitNoteDoc.submit();
+
+  t.equal('PINV-1003', debitNoteDoc.name, 'Debit Note name matches');
+
+  const debitNoteAles = await fyo.db.getAllRaw(
+    ModelNameEnum.AccountingLedgerEntry,
+    {
+      fields: ['name', 'account', 'credit', 'debit'],
+      filters: { referenceName: debitNoteDoc.name! },
+    }
+  );
+
+  for (const ale of debitNoteAles) {
+    if (ale.account === 'Stock Received But Not Billed') {
+      t.equal(
+        ale.debit,
+        '0.00000000000',
```

Comment:
Convert it to float and check, don't check against the string, if the precision changes, all these tests fail.